spec/unit/pops/types/p_object_type_spec.rb

```diff
@@ -1136,6 +1140,282 @@ def compile(code)
           "}]"
         ])
     end
+
+    context 'object with type parameters' do
+      it 'can be declared' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(['ok'])
+        type MyType = Object[
+          type_parameters => {
+            p1 => String
+          }]
+        notice('ok')
+        PUPPET
+      end
+
+      it 'can be referenced' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(["MyType['hello']"])
+        type MyType = Object[
+          type_parameters => {
+            p1 => String
+          }]
+
+        notice(MyType['hello'])
+        PUPPET
+      end
+
+      it 'can be declared with defaults' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(["MyType[default, 'world']"])
+        type MyType = Object[
+          type_parameters => {
+            p1 => { type => String, value => 'hello' },
+            p2 => String
+          }]
+
+        notice(MyType[default, 'world'])
+        PUPPET
+      end
+
+      it 'trailing defaults are skipped in string representation' do
+        expect(eval_and_collect_notices(<<-PUPPET, node)).to eql(["MyType['my']"])
```

Comment:
This is surprising when the given values happen to be the default values. I may not know what the defaults are, or may be writing code against two versions where the param is default in one version but not in another. Totally fine to drop defaults when they are give as `default`, but here it is a bit surprising. Is this difficult to deal with?

Reply:
Yes, it's difficult to deal with since the type then must remember if it was created using a default value or without a value. It would also mean that we get different String representation of two values that are exactly equal in every respect except the way they were initially declared. I'm against that because I'm a big fan of canonical and unambiguous presentation (and storage). The fact that the something can be declared with explicitly expressed defaults shouldn't affect the result.

Reply:
May need to reconsider that no given constraints always means "all instances". If that instead means with default constraints, then you need to use `Any` (etc) to relax the constraints. (Compare Hash that used to mean hash with Data). Alternatively - defaults for type parameters cannot pose additional constraints to the base type.

Reply:
"defaults for type parameters cannot pose additional constraints to the base type" sounds pretty much like what the current implementation does. I.e. the defaults have no effect whatsoever on the base type.

Reply:
What I meat is that a type T's default parameters must have this be true: `T[default, default, default,...] == T`